### PR TITLE
Autorun on OS startup

### PR DESCRIPTION
**Pull request by ghost**
Adds autorun on OS startup behaviour when "Start in tray" option is checked. Supported platforms are windows and xfreedesktop-compatible unixes. It works with multiple profiles although after creating/switching profile checkbox in settings is not updated. I am unsure how to go about it. Maybe it is another problem of not updating settings UI after switching profile?

OSX support is missing - someone please contribute. I added warning about missing implementation [here](https://github.com/novist/qTox/compare/tux3:master...novist:autorun?expand=1#diff-a095c041ed4fe02895e04ee7e2ddff6fR473). When it is implemented entire #ifdef / #endif block should be removed. With these changes behaviour on OSX does not change.

I hijacked "Start in tray" option because in code its called like get**Autostart**InTray(). This implied intention to use that option for autorun functionality in the future and so i did.

Windows implementation used std::wstring / QString::toStdWString(). While most likely it is not a problem im mentioning it here in case it is required to work with Qt compiled without stl support. Non-stl solution would be rather ugly so i opted in for tiny bit stl.

### CODE REVIEW

File: src/platform/autorun_win.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    inline std::wstring currentRegistryKeyName()
+    {
+        return (QString("qTox - ") + Settings::getInstance().getCurrentProfile()).toStdWString();
+    }
+}
+
+bool Platform::setAutorun(bool on)
+{
+    HKEY key = 0;
+    if(RegOpenKeyEx(HKEY_CURRENT_USER, L"Software\\Microsoft\\Windows\\CurrentVersion\\Run",
```
Comment (zetok) [inline src/platform/autorun_win.cpp]: According to qTox style, there should be space between `if` and `()`; `if()` → `if ()`
Could you please change that and other occurrences in code?

**Comment by dubslow** [comment]
I should think that starting qTox in tray and auto starting on boot are unrelated options, regardless of what the original author might have thought.

**Comment by ghost** [comment]
I would not agree. Think about it. What is desired when application runs on startup? That application runs in the background until we need it. What do we want when we start application manually? We want to interact with application immediately. Start in tray option would become obsolete because usually noone runs an application manually only to do nothing with it. I personally use "Start in Tray" option only with autorun feature (currently added manually to OS startup programs). Other tox instances have that option unchecked because i run them at will when i need them.

**Comment by Zer0-One** [comment]
"start in tray" and "run on boot" are unrelated options, and should have separate checkboxes.

**Comment by zetok** [comment]
@novist I have to disagree. I would prefer starting qTox in tray **not** when system starts if I knew that qTox will actually start. Currently I don't have this option on, because I don't feel like qTox is stable enough to be able to start properly without any sort of problem, and I need a visual confirmation that it actually has started.

Also, based on what has been said, and what little I could understand from code, you don't seem to consider scenario where user sets their os/de to start qTox with system, rather than checking box in qTox (no, _requiring_ user to do something through qTox when it can be done in different way is not a good thing)

**Comment by ProMcTagonist** [comment]
What Zetok said.

**Comment by ghost** [comment]
Well if you guys insist i can make another option but i really feel it would be a bloat ;)

**Comment by agilob** [comment]
I like when application starts on boot in tray and this feature should be supported in IM. I don't think it would be a bloat.

**Comment by ProMcTagonist** [comment]
As long as start in tray and start at boot are different checkboxes.

**Comment by dubslow** [comment]
Still need OS X implementation :P

**Comment by stqism** [comment]
@dubslow That's fairly easy to implement, going to need to install Qbloat first though.

**Comment by stqism** [comment]
@novist Fix your merge conflict, I cba to write both a launchagent and figure out how a bunch of Qt stuff works.

**Comment by ghost** [comment]
@stqism any update on this?
@dubslow i still think better having 66% implemented instead of 0%. Maybe this could be merged while osx support could be contributed via new PR? At the moment on mac checkbox is disabled so its graceful-enough failure. Or it can be hidden if you prefer.

**Comment by ProMcTagonist** [comment]
Start in Tray and Autostart are not the same.

**Comment by ghost** [comment]
That setting was split long ago, just like you wanted ;)

**Comment by ProMcTagonist** [comment]
<3

**Comment by stqism** [comment]
@novist Totally forgot

**Comment by stqism** [comment]
Yeah, I remember, I put that on hold because your pr didn't build with core and you didn't mention that it got fixed.

**Comment by ghost** [comment]
Dont new commits generate notification for people watching PR? I thought everyone gets notified auto. Hmm..

**Comment by stqism** [comment]
Nope, I just watch the repo and don't remember getting notified.

**Comment by zetok** [comment]
@novist nope, pushing new commits to PR doesn't generate notification

**Comment by ghost** [comment]
@stqism did you start any work based on this PR? Asking because i would fix merge conflicts rebasing changes on to upstream for nicer history. Otherwise if you are already working on stuff i would do merge commit.

**Comment by dubslow** [comment]
I am generally not a fan of rebasing

**Comment by ghost** [comment]
Clean > spaghetti history. One just have to know when to rebase and when to merge ;)

**Comment by ghost** [comment]
Solved conflicts. Since OSX support does not seem to be coming any time soon i suggest this to be merged. There simply is no reason to have windows/linux users waiting for OSX support. I know for a fact that it is major deficiency for non-expert windows users. Not everyone knows of startup folder in start menu. Worse yet - it is kind of gone from windows 8. I had to dig that up deep in %APPDATA% and toss qTox shortcut in there just so my family member would be online when he is online instead of being online whenever he remembers to start qTox. So if all good merge pl0x.

Also maybe we want this option to be enabled first time qTox is installed and run? I think its usually what majority would want.

**Comment by stqism** [comment]
Feel free, I need to finish my updater work, hook in to the notify api, and create a launchd service before it'll ever be start-able on boot. 

Thing to note though, we need to make it a part of the UI with some sort of getting started somehow. With Linux we can't just shove a check box in to the deb install.

**Comment by ghost** [comment]
If you mean autorun by default after install - well i was thinking it being automatic. Check if its first launch, if it is set autorun=true. As for more controlled approach - well windows installer could have autorun checkbox but i dont think it it would work. Problem i see here is using profile name in launch command line. Its not exactly known at install time. Besides its problematic to know if it is first install or not. If app is being updated and autostart was disabled by user we do not want to override this behaviour.

**Pull request merged**